pkgs/development/web/nodejs/default.nix

```diff
@@ -0,0 +1,555 @@
+          NINJA = "TERM=dumb ninja";
+        }
+        // lib.optionalAttrs (stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isx86_64) {
+          # Make sure libc++ uses `posix_memalign` instead of `aligned_alloc` on x86_64-darwin.
+          # Otherwise, nodejs would require the 11.0 SDK and macOS 10.15+.
+          NIX_CFLAGS_COMPILE = "-D__ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__=101300 -Wno-macro-redefined";
+        };
+
+      nativeBuildInputs =
+        [
+          installShellFiles
+          ninja
+          pkgconf
+          python
+        ]
+        ++ lib.optionals stdenv.buildPlatform.isDarwin [
+          # gyp checks `sysctl -n hw.memsize` if `sys.platform == "darwin"`.
+          unixtools.sysctl
+        ]
+        ++ lib.optionals stdenv.hostPlatform.isDarwin [
+          # For gyp-mac-tool if `flavor == "mac"`.
+          darwin-cctools-only-libtool
+        ];
+
+      buildInputs =
+        [
+          zlib
+          libuv
+          openssl
+          icu
+          # For patchShebangsAuto. Technically, we do not need bash in build
+          # inputs since all scripts are wrappers over the corresponding JS
+          # scripts. There are some packages though that use bash wrappers, e.g.
+          # polaris-web.
+          bash
+        ]
+        ++ lib.optionals stdenv.hostPlatform.isDarwin [
+          CoreServices
+          ApplicationServices
+        ];
+
+      # We currently rely on Makefile and stdenv for build phases, so do not let
+      # ninja’s setup hook to override default stdenv phases.
+      dontUseNinjaBuild = true;
+      dontUseNinjaCheck = true;
+
+      # Configure script is compatible with autoconf-style output flags.
+      setOutputFlags = false;
+
+      configureFlags =
+        [
+          "--prefix=" # see patches/configure-remove-expanduser.patch
+          "--ninja"
+          "--shared-openssl"
+          "--shared-zlib"
+          "--shared-libuv"
+          "--with-intl=system-icu"
+          "--openssl-use-def-ca-store"
+          "--no-cross-compiling"
+          "--dest-os=${destOS}"
+          "--dest-cpu=${destCPU}"
+        ]
+        ++ lib.optionals (destARMFPU != null) [ "--with-arm-fpu=${destARMFPU}" ]
+        ++ lib.optionals (destARMFloatABI != null) [ "--with-arm-float-abi=${destARMFloatABI}" ]
+        ++ lib.optionals (!canExecute) [
+          # Node.js requires matching bitness between build and host platforms, e.g.
+          # for V8 startup snapshot builder (see tools/snapshot) and some other
+          # tools. We apply a patch that runs these tools using a host platform
+          # emulator and avoid cross-compiling altogether (from the build system’s
+          # perspective).
+          "--emulator=${emulator}"
+        ]
+        ++ lib.optionals (lib.versionOlder version "19") [ "--without-dtrace" ]
+        ++ lib.optionals finalAttrs.slim [
+          "--without-npm"
+          "--without-corepack"
+        ];
```

Comment:
Not sure about this change, I've been using the `-slim` versions and been using Corepack to let me download package managers. If we remove Corepack, that would be quite a breaking change for me.

Reply:
I'm not necessarily against removing corepack but is `slim` really a good argument name? I like the more descriptive `withNpm` and we could add `withCorepack`.

Reply:
I don’t think slim versions were supposed to contain any bundled node_modules, be that npm or corepack. Corepack was added in v16.9.0, v14.19.0 (https://nodejs.org/api/corepack.html) and IIRC nodejs-slim was introduced before that.

>I've been using the -slim versions and been using Corepack to let me download package managers. If we remove Corepack, that would be quite a breaking change for me.

Is there any reason for using nodejs-slim over normal nodejs package? There shouldn’t be any difference between them aside from missing npm. We also have separate corepack package in Nixpkgs that IIRC you’ve reviewed some time ago.

Reply:
>is slim really a good argument name? I like the more descriptive withNpm and we could add withCorepack.

It is not 😅
I’ll change it back to `enableNpm` with additional `enableCorepack`.

Reply:
I think we would not want to add another variant built by Hydra for every Node version, so we would still need to decide what “slim” means, and in that case I’m not sure there’s a benefit to having the additional flag.

Is there any chance that we could simply separate out npm and Corepack into separate derivations so they don’t need to rebuild the whole Node.js?

Reply:
I think it's fine to have `-slim` mean _without npm and Corepack_, but that should probably not be backported to 24.05, as I'd expect it to be breaking to more folks than just me.

Yeah sure finding a workaround (either switching to non-slim package, or install Corepack separately) is probably trivial given Nix user-base, and to clarify I'm not against this change for the long term / new packages.

Reply:
I think we can, although I’d expect a lot of changes for that because some packages (as well as Node.js build infrastructure in Nixpkgs) may assume that `npm` is provided by `nodejs` package (Node.js bundles Corepack and NPM, including all required dependencies, e.g. node-gyp). That said, it’d be nice to bootstrap a separate `npm` package using bundled npm from Node.js source distribution.

Reply:
I assume Debian does that because they build Node.js with `--without-{npm,corepack}`.
https://salsa.debian.org/js-team/nodejs/-/blob/94873a3915ac88b3f423c44a44e90db939a151ad/debian/rules#L26-27

Reply:
We could have `nodejs` be a wrapper around `nodejs-unwrapped` + `npm` + `corepack` or something.

Reply:
> We could have `nodejs` be a wrapper around `nodejs-slim` + `npm` + `corepack` or something, so that the dependencies still work out.

What's the advantage compared to adding `npm` to a few packages? I doubt it's even that many because `buildNpmPackage` would obviously still come with `npm` by default and other packages tend to use yarn or pnpm.

Reply:
Well, more to the point users have `nodejs` installed in their environments and expect to have `npm`.